### PR TITLE
Work in progress - data adapter, chat bot

### PR DESCRIPTION
**Pull request by MartynaUb**
Do not merge yet

### CODE REVIEW

File: Support_Your_Locals/Controllers/AdminController.cs [hunk 1]
```diff
@@ -15,38 +14,50 @@ namespace Support_Your_Locals.Controllers
...
+            repository = repo;
+
         }
         [HttpGet]
         [Authorize(Roles = "Administrator")]
         public ActionResult Index()
         {
+            var businesses = repository.GetBusinesses();
+            var usersBusinesses = businesses.GroupBy(business => business.User);
+            var users = repository.GetUsers().Where(notAdmin);
```
Comment (EvalVis) [inline Support_Your_Locals/Controllers/AdminController.cs:36]: From where is this not admin coming? Are admins in the same schema as standart users?

File: Support_Your_Locals/Models/Repositories/LegacyServiceRepository.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            {
+                ProductID = row.Field<long>("ProductID"),
+                Name = row.Field<string>("Name"),
+                PricePerUnit = row.Field<decimal>("PricePerUnit"),
+                Unit = row.Field<string>("Unit"),
+                Comment = row.Field<string>("Comment"),
+                Picture = row.Field<string>("Picture"),
+                BusinessID = row.Field<long>("BusinessID"),
+            }).ToList();
+        }
```
Comment (EvalVis) [inline Support_Your_Locals/Models/Repositories/LegacyServiceRepository.cs:171]: Edit business missing.

File: Support_Your_Locals/Controllers/AdminController.cs (hunk 1 above)
Comment (Yar3k) [reply Support_Your_Locals/Controllers/AdminController.cs:36]: Probably in future there will be added roles, so if some kind of person will have "Administrator" role in cookie then he will have access to this page

**Review by Yar3k — APPROVED** [review]

File: Support_Your_Locals/Controllers/AdminController.cs (hunk 1 above)
Comment (MartynaUb) [reply Support_Your_Locals/Controllers/AdminController.cs:36]: We already have roles in our authentication mechanism. That's why there is `[Authorize(Roles = "Administrator")]` so only users with `Administrator` role can access this method. 
`isAdmin` is predicate function to filter out users that are not admins from user list since they are stored in the same schema. I think admin users are not relevant in the admin page that is why I filter them out

**Comment by MartynaUb** [comment]
Lots of files changed with this pull request. Would recommend reading it commit by commit to minimize the confusion

File: Support_Your_Locals/Controllers/AdminController.cs (hunk 1 above)
Comment (EvalVis) [reply Support_Your_Locals/Controllers/AdminController.cs:36]: That explains it. Thanks.

File: Support_Your_Locals/Controllers/AdminController.cs [hunk 3]
```diff
@@ -15,38 +14,50 @@ namespace Support_Your_Locals.Controllers
...
         public ActionResult DeleteUser(long id)
         {
-            repository.DeleteUser(new Models.User { UserID = id });
+            repository.DeleteUser(id);
+            return RedirectToAction("Index");
+        }
+
+        [Authorize(Roles = "Administrator")]
+        public ActionResult Reasign(long id, long targetUserId)
+        {
```
Comment (EvalVis) [inline Support_Your_Locals/Controllers/AdminController.cs:78]: I do not actually understand what reassign does. It just redirects. Is this part of WIP?

File: Support_Your_Locals/Models/Repositories/LegacyServiceRepository.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+namespace Support_Your_Locals.Models.Repositories
+{
+    public class LegacyServiceRepository : ILegacyServiceRepository
+    {
+
+        private String connectionString;
+        
+        public LegacyServiceRepository(IConfiguration configuration)
+        {
+            this.connectionString = configuration["ConnectionStrings:DefaultConnection"];
```
Comment (EvalVis) [inline Support_Your_Locals/Models/Repositories/LegacyServiceRepository.cs:17]: This could be simplified to `Configuration.GetConnectionString("DefaultConnection")`.

File: Support_Your_Locals/Views/Admin/Index.cshtml [hunk 5]
```diff
@@ -49,9 +49,11 @@
             </tr>
         </thead>
         <tbody id="users-table" style="display:none;">
-            @foreach (var user in Model.Users)
+            @foreach (var userBusinesses in Model.UsersBusinesses)
             {
-                <tr>
+                var @user = userBusinesses.Key;
+                <tr style="color: white; background-color: #27303d">
+                    <td onclick="enableMultiple(@Html.Raw(Json.Serialize(@userBusinesses.Select(business => business.BusinessID))))" style="cursor:pointer">List businesses</td>
```
Comment (EvalVis) [inline Support_Your_Locals/Views/Admin/Index.cshtml:56]: You could transfer some of this code to `@{
}` section, I think.

File: Support_Your_Locals/Controllers/AdminController.cs [hunk 6]
```diff
@@ -15,38 +14,50 @@ namespace Support_Your_Locals.Controllers
...
             configuration = config;
+            repository = repo;
+
         }
         [HttpGet]
         [Authorize(Roles = "Administrator")]
         public ActionResult Index()
         {
+            var businesses = repository.GetBusinesses();
+            var usersBusinesses = businesses.GroupBy(business => business.User);
```
Comment (EvalVis) [inline Support_Your_Locals/Controllers/AdminController.cs:34]: This is a good solution, since I don't think ADO.NET has Navigation features like .Include()

File: Support_Your_Locals/Models/User.cs [hunk 7]
```diff
@@ -13,5 +13,15 @@ public class User
         public string Passhash { get; set; }
         public List<Business> Businesses { get; set; } = new List<Business>();
 
+        public override bool Equals(object obj)
+        {
+            return obj is User user &&
+                   UserID == user.UserID;
+        }
+
```
Comment (EvalVis) [inline Support_Your_Locals/Models/User.cs:21]: I do not see Equals used anywhere now. Is this a preparation for the future?

File: Support_Your_Locals/Controllers/AdminController.cs (hunk 3 above)
Comment (EvalVis) [reply Support_Your_Locals/Controllers/AdminController.cs:78]: OK, I see it was replaced with AnswerQuestion in the next commit.

File: Support_Your_Locals/Controllers/AdminController.cs [hunk 8]
```diff
@@ -56,8 +61,23 @@ public ActionResult DeleteUser(long id)
...
+                Email = email,
+            });
+        }
+
+
+        [Authorize(Roles = "Administrator")]
+        [HttpPost]
+        public ActionResult AnswerQuestion(AnswerQuestionViewModel model)
         {
+            ResponseEvent(this, new ResponseEventArgs(model.Email, model.AnwserText));
```
Comment (EvalVis) [inline Support_Your_Locals/Controllers/AdminController.cs]: Resharper indicates possible null reference exception. Try ` ResponseEvent?.Invoke(this, new ResponseEventArgs(model.Email, model.AnwserText));`.

**Comment by EvalVis** [comment]
Checkout commit 1bf4b396da87f9b29f318d1b976efafd9314ed63. It contains a small fix.

File: Support_Your_Locals/Controllers/QuestionRestController.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        private ILegacyServiceRepository repository;
+
+        public QuestionRestController(ILegacyServiceRepository repo)
+        {
+            repository = repo;
+
+        }
+       
+        [HttpPost]
+        public IActionResult Post([FromBody] AddQuestionResource question)
```
Comment (EvalVis) [inline Support_Your_Locals/Controllers/QuestionRestController.cs:22]: Since you decorated the controller class with `[ApiController]` you do not need to check the ModelState. This will be done automatically and 400 Bad Request error code will be sent if Model is invalid. You also do not need [From Body]
[ApiController] does that for you as well.

File: Support_Your_Locals/Models/Question.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿namespace Support_Your_Locals.Models
+{
+    public class Question
+    {
+        public long QuestionId { get; set; }
+        public string Email { get; set; }
+        public string Text { get; set; }
+        public string Response { get; set; }
```
Comment (EvalVis) [inline Support_Your_Locals/Models/Question.cs:8]: So, if the question was not answered in the first time, another answer (response) will overwrite the first one, right?

Comment (EvalVis) [reply Support_Your_Locals/Models/Question.cs:8]: OK, I now understand. Then we response is said, the question is answered. Techically, you could just check Response == null, but this will also fit.

File: Support_Your_Locals/Views/Admin/Index.cshtml [hunk 11]
```diff
@@ -74,6 +74,43 @@
...
+            {
+            <tr>
+                <td>@question.Email</td>
+                <td>@question.Text</td>
+                <td>@(question.IsAnswered ? "Answered" : "Not answered")</td>
+                <td>
+                    @if (!question.IsAnswered)
+                    {
+
+                        @Html.ActionLink("Answer", "AnswerQuestion", new { id = @question.QuestionId, email = question.Email })
```
Comment (EvalVis) [inline Support_Your_Locals/Views/Admin/Index.cshtml:100]: This could be done more elegantly with anchor tag helpers `<a asp-action="AnswerQuestion" asp-route-id="@question.QuestionId" asp-route-email="@question.Email">`.

Comment (EvalVis) [reply Support_Your_Locals/Views/Admin/Index.cshtml:100]: You can try if you like this.

File: LocalsTradeBot/LocalsTradeBot/AdapterWithErrorHandler.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        public AdapterWithErrorHandler(IConfiguration configuration, ILogger<BotFrameworkHttpAdapter> logger)
+            : base(configuration, logger)
+        {
+            OnTurnError = async (turnContext, exception) =>
+            {
+                // Log any leaked exception from the application.
+                logger.LogError(exception, $"[OnTurnError] unhandled error : {exception.Message}");
+
+                // Send a message to the user
+                await turnContext.SendActivityAsync("The bot encountered an error or bug.");
```
Comment (EvalVis) [inline LocalsTradeBot/LocalsTradeBot/AdapterWithErrorHandler.cs:24]: OK, I am not sure if we should ask a user to fix the bot.

File: LocalsTradeBot/LocalsTradeBot/Bots/EchoBot.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            }
+
+            public CurrentConversationState State { get; set; } = CurrentConversationState.Question;
+        }
+
+        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
+        {
+
+            var conversationStateAccessors = _conversationState.CreateProperty<Conversation>(nameof(Conversation));
+            var conversation = await conversationStateAccessors.GetAsync(turnContext, () => new Conversation(), cancellationToken);
```
Comment (EvalVis) [inline LocalsTradeBot/LocalsTradeBot/Bots/EchoBot.cs:39]: OK, definitely, I do not understand, what this does.

File: LocalsTradeBot/LocalsTradeBot/Controllers/BotController.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IBotFrameworkHttpAdapter Adapter;
+        private readonly IBot Bot;
+
+        public BotController(IBotFrameworkHttpAdapter adapter, IBot bot)
+        {
+            Adapter = adapter;
+            Bot = bot;
+        }
+
+        [HttpPost, HttpGet]
```
Comment (EvalVis) [inline LocalsTradeBot/LocalsTradeBot/Controllers/BotController.cs:29]: Why HttpGet on Post?

File: LocalsTradeBot/LocalsTradeBot/DeploymentTemplates/template-with-new-rg.json [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+                                    },
+                                    "webSocketsEnabled": true
+                                }
+                            }
+                        },
+                        {
+                            "apiVersion": "2017-12-01",
+                            "type": "Microsoft.BotService/botServices",
+                            "name": "[parameters('botId')]",
+                            "location": "global",
```
Comment (EvalVis) [inline LocalsTradeBot/LocalsTradeBot/DeploymentTemplates/template-with-new-rg.json:159]: Is this file relevant to us?

File: LocalsTradeBot/LocalsTradeBot/DeploymentTemplates/template-with-new-rg.json [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+                            "kind": "bot",
+                            "sku": {
+                                "name": "[parameters('botSku')]"
+                            },
+                            "properties": {
+                                "name": "[parameters('botId')]",
+                                "displayName": "[parameters('botId')]",
+                                "endpoint": "[variables('botEndpoint')]",
```
Comment (EvalVis) [inline LocalsTradeBot/LocalsTradeBot/DeploymentTemplates/template-with-new-rg.json:167]: This could be relevant.

**Comment by EvalVis** [comment]
OK, I just ran bot emulator and entered Yes after "Hello! Do you have any questions you Want to ask?"

**Comment by EvalVis** [comment]
We could write: "... if so, please ask right away."

**Comment by EvalVis** [comment]
Also, the bot does not know if the information was delivered or not. We will probably want to fix this.

**Comment by EvalVis** [comment]
Since Support Your Locals now has API Controllers, I think swagger will fit.

**Review by EvalVis — APPROVED** [review]
Well done.

**Comment by MartynaUb** [comment]
> We could write: "... if so, please ask right away."

Good idea, that would be clearer
